### ISSUE TITLE
Still getting issue: 2 keys issued for same client. Follow up to #414

### ISSUE THREAD

**Issue by deanclkclk** [spring-projects/spring-security-oauth#463]
OK, I'm still getting the same issue after following the suggestion on stackoverflow:

http://stackoverflow.com/questions/28730821/spring-security-fails-jdbcclientdetailsservice-error/28737845?noredirect=1#comment45781580_28737845

     title>Error 500 Incorrect result size: expected 1, actual 2</title>
     </head>
     <body><h2>HTTP ERROR 500</h2>
     <p>Problem accessing /oauth/token. Reason:
     <pre>    Incorrect result size: expected 1, actual 2</pre></p><h3>Caused by:</h3>      <pre>org.springframework.dao.IncorrectResultSizeDataAccessException: Incorrect result size:  expected 1, actual 2
 	at   org.springframework.dao.support.DataAccessUtils.requiredSingleResult(DataAccessUtils.java:74)
	at org.springframework.jdbc.core.JdbcTemplate.queryForObject(JdbcTemplate.java:812)
	at org.springframework.security.oauth2.provider.token.JdbcTokenStore.getAccessToken(JdbcTokenStore.java:105)
	at org.springframework.security.oauth2.provider.token.DefaultTokenServices.createAccessToken(DefaultTokenServices.java:75)
	at org.springframework.security.oauth2.provider.token.AbstractTokenGranter.getAccessToken(AbstractTokenGranter.java:68)
	at org.springframework.security.oauth2.provider.token.AbstractTokenGranter.grant(AbstractTokenGranter.java:60)
	at org.springframework.security.oauth2.provider.client.ClientCredentialsTokenGranter.grant(ClientCredentialsTokenGranter.java:41)
	at org.springframework.security.oauth2.provider.CompositeTokenGranter.grant(CompositeTokenGranter.java:38)
	at org.springframework.security.oauth2.provider.endpoint.TokenEndpoint.getAccessToken(TokenEndpoint.java:100)
	at sun.reflect.GeneratedMethodAccessor241.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.web.method.support.InvocableHandlerMethod.invoke(InvocableHandlerMethod.java:215)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:132)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:104)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:749)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:689)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:83)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:938)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:870)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:961)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:863)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:727)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:837)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:538)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1352)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:118)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:84)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:113)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:154)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.authentication.www.BasicAuthenticationFilter.doFilter(BasicAuthenticationFilter.java:150)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.oauth2.provider.client.ClientCredentialsTokenEndpointFilter.successfulAuthentication(ClientCredentialsTokenEndpointFilter.java:117)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:235)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:87)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:342)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:192)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:160)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:344)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:261)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1323)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:476)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:119)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:517)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:225)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:937)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:406)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:183)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:871)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:117)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:247)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:149)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:110)
	at org.eclipse.jetty.server.Server.handle(Server.java:346)
	at org.eclipse.jetty.server.HttpConnection.handleRequest(HttpConnection.java:589)
	at org.eclipse.jetty.server.HttpConnection$RequestHandler.content(HttpConnection.java:1065)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:823)
	at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:220)
	at org.eclipse.jetty.server.HttpConnection.handle(HttpConnection.java:411)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:535)
	at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:40)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:529)
	at java.lang.Thread.run(Thread.java:745)

And my security XML:

     <http pattern="/oauth/token" create-session="stateless"   authentication-manager-ref="authenticationManager" xmlns="http://www.springframework.org/schema/security">
		<intercept-url pattern="/oauth/token" access="ROLE_FRONTEND"  />
		<anonymous enabled="false" />
		<http-basic entry-point-ref="clientAuthenticationEntryPoint"  />
		
		<custom-filter ref="clientCredentialsTokenEndpointFilter" before="BASIC_AUTH_FILTER" />
		
		<access-denied-handler ref="oauthAccessDeniedHandler" />
	</http>
	
	
	<http pattern="/public/person" create-session="stateless" entry-point-ref="oauthAuthenticationEntryPoint"
		access-decision-manager-ref="accessDecisionManager" xmlns="http://www.springframework.org/schema/security">
		<anonymous enabled="false" />
		<intercept-url pattern="/public/person/credential_verfiy" access="ROLE_FRONTEND"  />
		<custom-filter ref="resourceServerFilter" before="PRE_AUTH_FILTER" />
		<access-denied-handler ref="oauthAccessDeniedHandler" />
	</http>
	
	
	<http pattern="/public/registration" create-session="stateless" entry-point-ref="oauthAuthenticationEntryPoint"
		access-decision-manager-ref="accessDecisionManager" xmlns="http://www.springframework.org/schema/security">
		<anonymous enabled="false" />
		<intercept-url pattern="/public/registration/activation/**" access="IS_AUTHENTICATED_ANONYMOUSLY" />
		<intercept-url pattern="/public/registration" access="ROLE_FRONTEND"  />
		<custom-filter ref="resourceServerFilter" before="PRE_AUTH_FILTER" />
		<access-denied-handler ref="oauthAccessDeniedHandler" />
	</http>

	<http pattern="/private/vote" create-session="stateless" entry-point-ref="oauthAuthenticationEntryPoint"
		  access-decision-manager-ref="accessDecisionManager" xmlns="http://www.springframework.org/schema/security">
		<anonymous enabled="false" />
		<intercept-url pattern="/private/vote" access="ROLE_FRONTEND"  />
		<custom-filter ref="resourceServerFilter" before="PRE_AUTH_FILTER" />
		<access-denied-handler ref="oauthAccessDeniedHandler" />
	</http>

	<http pattern="/private/currencies" create-session="stateless" entry-point-ref="oauthAuthenticationEntryPoint"
		  access-decision-manager-ref="accessDecisionManager" xmlns="http://www.springframework.org/schema/security">
		<anonymous enabled="false" />
		<intercept-url pattern="/private/currencies" access="ROLE_FRONTEND"  />
		<custom-filter ref="resourceServerFilter" before="PRE_AUTH_FILTER" />
		<access-denied-handler ref="oauthAccessDeniedHandler" />
	</http>
	
	<authentication-manager alias="authenticationManager"  xmlns="http://www.springframework.org/schema/security">
		<sec:authentication-provider user-service-ref="clientDetailsUserService" />
	</authentication-manager>

	<authentication-manager id="authenticationManager2"  xmlns="http://www.springframework.org/schema/security">
		<sec:authentication-provider ref="customDaoProvider">
		</sec:authentication-provider>
	</authentication-manager>
	
		
	<beans:bean id="customUserDetailService" class="com.cointraders.api.services.UserDetailsServiceImpl">
		<beans:property name="accountService" ref="userAccountServiceImpl"></beans:property>
		<beans:property name="personService" ref="personServiceImpl"></beans:property>
	</beans:bean>
	
	
	<beans:bean id="customDaoProvider" class="com.cointraders.api.daoauthproviders.CustomDaoAuthenticationProvider">
        <beans:property name="userDetailsService" ref="customUserDetailService"/>
        <beans:property name="passwordEncoder" ref="passwordEncoder" />
        <beans:property name="accountService" ref="userAccountServiceImpl"></beans:property>
		<beans:property name="clientService" ref="clientServiceImpl"></beans:property>
    </beans:bean>
    

	<beans:bean id="clientDetails" class="org.springframework.security.oauth2.provider.JdbcClientDetailsService">
	    <beans:constructor-arg ref="dataSource" />
	</beans:bean>
	
	<beans:bean id="accessDecisionManager" class="org.springframework.security.access.vote.UnanimousBased" xmlns="http://www.springframework.org/schema/beans">
		<beans:constructor-arg>
			<beans:list>
				<beans:bean class="org.springframework.security.oauth2.provider.vote.ScopeVoter" />
				<beans:bean class="org.springframework.security.access.vote.RoleVoter" />
				<beans:bean class="org.springframework.security.access.vote.AuthenticatedVoter" />
			</beans:list>
		</beans:constructor-arg>
	</beans:bean>	
	
	<oauth:authorization-server  client-details-service-ref="clientDetails" token-services-ref="tokenServices">
		<oauth:refresh-token />
		<oauth:client-credentials/>
		<oauth:custom-grant token-granter-ref="customPasswordGrant"/>
		<oauth:custom-grant token-granter-ref="randomTokenGrant" />
	</oauth:authorization-server>
	
	<beans:bean id="restServicesSuccessHandler" class="com.cointraders.api.handlers.RestAuthenticationSuccessHandler" />	

	
	<oauth:resource-server id="resourceServerFilter" resource-id="api" token-services-ref="tokenServices" />
	
	<sec:global-method-security pre-post-annotations="enabled" proxy-target-class="true">
		<sec:expression-handler ref="oauthExpressionHandler" />
	</sec:global-method-security>

	<oauth:expression-handler id="oauthExpressionHandler" />

	<oauth:web-expression-handler id="oauthWebExpressionHandler" />
	
	<beans:bean id="clientAuthenticationEntryPoint" class="org.springframework.security.oauth2.provider.error.OAuth2AuthenticationEntryPoint">
		<beans:property name="realmName" value="api/" />
	</beans:bean>
	
	
	<beans:bean id="randomTokenGrant" class="com.cointraders.api.grants.RandomTokenGrant">
		<beans:constructor-arg index="0"><beans:value>${api.common.login.oauth.token}</beans:value></beans:constructor-arg>
		<beans:constructor-arg index="1"><beans:value>${user.login.username.param}</beans:value></beans:constructor-arg>
		<beans:constructor-arg index="2"><beans:value>${user.login.password.param}</beans:value></beans:constructor-arg>
		<beans:constructor-arg index="3" ref="authenticationManager2"></beans:constructor-arg>
		<beans:constructor-arg index="4" ref="tokenServices"></beans:constructor-arg>
		<beans:constructor-arg index="5" ref="clientDetails"></beans:constructor-arg>
		<beans:constructor-arg index="6"><beans:value>${api.common.login.oauth.grant_type.two_factor}</beans:value></beans:constructor-arg>
		<beans:constructor-arg index="7" ref="userAccountServiceImpl"></beans:constructor-arg>
	</beans:bean>
	
	<beans:bean id="customPasswordGrant" class="com.cointraders.api.grants.CustomPasswordGrant">
		<beans:constructor-arg index="0" ref="authenticationManager2"></beans:constructor-arg>
		<beans:constructor-arg index="1" ref="tokenServices"></beans:constructor-arg>
		<beans:constructor-arg index="2" ref="clientDetails"></beans:constructor-arg>
		<beans:constructor-arg index="3" ref="userAccountServiceImpl"></beans:constructor-arg>
		<beans:constructor-arg index="4"><beans:value>${api.common.login.oauth.grant_type.custom_password}</beans:value></beans:constructor-arg>
	</beans:bean>
	
	
	
	<beans:bean id="clientCredentialsTokenEndpointFilter" class="org.springframework.security.oauth2.provider.client.ClientCredentialsTokenEndpointFilter">
		<beans:property name="authenticationManager" ref="authenticationManager" />
	</beans:bean>
	
	
	<beans:bean id="tokenStore" class="org.springframework.security.oauth2.provider.token.JdbcTokenStore">
	    <beans:constructor-arg ref="dataSource" />
	</beans:bean>
	<beans:bean id="tokenServices" class="org.springframework.security.oauth2.provider.token.DefaultTokenServices">
		<beans:property name="tokenStore" ref="tokenStore" />
		<beans:property name="supportRefreshToken" value="true" />
		<beans:property name="clientDetailsService" ref="clientDetails" />
	</beans:bean>
	
	<beans:bean id="clientDetailsUserService" class="org.springframework.security.oauth2.provider.client.ClientDetailsUserDetailsService">
		<beans:constructor-arg ref="clientDetails" />
	</beans:bean>
	<beans:bean id="oauthAccessDeniedHandler" class="org.springframework.security.oauth2.provider.error.OAuth2AccessDeniedHandler" />
	
	<beans:bean id="oauthAuthenticationEntryPoint" class="org.springframework.security.oauth2.provider.error.OAuth2AuthenticationEntryPoint">
		<beans:property name="realmName" value="api" />
	</beans:bean>

	<beans:bean id="clientServiceImpl" class="com.cointraders.api.services.ClientServiceImpl">
		<beans:property name="clientDetailsService" ref="clientDetails" />
	</beans:bean>

**Comment by derkoe** [comment]
There is a race condition in `org.springframework.security.oauth2.provider.token.DefaultTokenServices#refreshAccessToken` which results in entries with duplicate `authentication_id` in the table `oauth_access_token`. This occurs when two threads access the database at the same time but both still get the authentication for the old refresh-token (`tokenStore.readAuthenticationForRefreshToken(refreshToken)`).

You can workaround this making `authentication_id` unique. Then the second attempt to refresh the token fails with a DuplicateKeyException which you can handle. This is not that bad since duplicate refresh requests aren't that good anyway.

The schema of the OAuth2 test sets the authentication_id as primary key (which is always unique): spring-security-oauth2/src/test/resources/schema.sql 

    create table oauth_access_token (
      token_id VARCHAR(256),
      token LONGVARBINARY,
      authentication_id VARCHAR(256) PRIMARY KEY,
      user_name VARCHAR(256),
      client_id VARCHAR(256),
      authentication LONGVARBINARY,
      refresh_token VARCHAR(256)
    );

**Comment by dsyer** [comment]
That's not really a workaround, that's just the correct way to set up the schema.

**Comment by xenoterracide** [comment]
Can we re-open this please, to add a decent exception  on unique constraint failure, or maybe have smarter transactional code that first finds out if it exists instead of doing blind inserts. Also #1200 this "correct behavior" results in information exposure if not otherwise handled


```
parallel --jobs 4 -X 'curl -X POST --header "Content-Type: application/x-www-form-urlencoded" --header "Accept: application/json" -d "grant_type=password&client_id=O88trBJRSemI92Lu_CuC3Q&username=ou9xput8rraa7cedkdohxa&password=sna1C8jSSOK3dAiurGXQTA" --silent http://localhost:8080/oauth/token' ::: echo {1..20}
{"access_token":"edb66ad0-7e4f-4481-8e8a-fb75d747047d","token_type":"bearer","expires_in":599,"scope":"read","client_id":"O88trBJRSemI92Lu_CuC3Q"}{"access_token":"edb66ad0-7e4f-4481-8e8a-fb75d747047d","token_type":"bearer","expires_in":598,"scope":"read","client_id":"O88trBJRSemI92Lu_CuC3Q"}{"error":"server_error","error_description":"PreparedStatementCallback; SQL [insert into oauth_access_token (token_id, token, authentication_id, user_name, client_id, authentication, refresh_token) values (?, ?, ?, ?, ?, ?, ?)]; ORA-00001: unique constraint (DEX_DEV_USER.PK_OAT_AID) violated\n; nested exception is java.sql.SQLIntegrityConstraintViolationException: ORA-00001: unique constraint (DEX_DEV_USER.PK_OAT_AID) violated\n"}{"error":"server_error","error_description":"PreparedStatementCallback; SQL [insert into oauth_access_token (token_id, token, authentication_id, user_name, client_id, authentication, refresh_token) values (?, ?, ?, ?, ?, ?, ?)]; ORA-00001: unique constraint (DEX_DEV_USER.PK_OAT_AID) violated\n; nested exception is java.sql.SQLIntegrityConstraintViolationException: ORA-00001: unique constraint (DEX_DEV_USER.PK_OAT_AID) violated\n"}```

**Comment by dsyer** [comment]
You can create a new issue if you like and explain why you didn’t set up the schema properly. I don’t like to re-open and issue that is already closed (nothing changed).

**Comment by xenoterracide** [comment]
did you read the above code that was pasted? we've corrected our schema per this, but the system still doesn't consistently handle the failure properly. I see about a 10% error rate on our localhost with some exceptions doing 100 requests, 10 concurrently. Now I can make another ticket if you really want, obviously I've already made one regarding the information exposure, but this one doesn't seem to truly handle the issue as is, 'because the fix, or lack thereof, still results in concurrency failure.
